File: paperclip/ryvie-accounts.mjs
/**
 * ryvie-accounts.mjs — Fiche de gestion des comptes Paperclip.
 *
 * Vit AVEC l'app (appstore) et s'exécute DANS le conteneur app-paperclip-server,
 * qui embarque déjà better-auth + le driver pg. Ryvie ne fait que la déclencher
 * (docker exec) et lire son résultat ; il ne connaît ni better-auth ni le schéma.
 *
 * Paperclip s'authentifie via better-auth : le hash du mot de passe est en
 * scrypt (format `salt:clé` hexa, PAS bcrypt), stocké dans la table `account`
 * (colonnes snake_case : user_id, provider_id='credential', password). On
 * réutilise donc le hasher natif de better-auth pour garantir la compatibilité.
 *
 * Sous-commandes :
 *   list                       → JSON des comptes sur stdout
 *   reset  (env RESET_ID/PWD)  → réécrit le hash + vérifie → "OK" / "FAIL"
 */
import { createRequire } from 'node:module';
import { hashPassword, verifyPassword } from 'better-auth/crypto';

// pg est un dep transitif (pnpm) : on le résout sans dépendre d'un symlink racine.
const require = createRequire(import.meta.url);
const pgPath = require.resolve('pg', {
  paths: [
    '/app/packages/db/node_modules',
    '/app/server/node_modules',
    '/app/node_modules/.pnpm/pg@8.18.0/node_modules',
  ],
});
const { Client } = require(pgPath);

const client = new Client({ connectionString: process.env.DATABASE_URL });
await client.connect();

try {
  const cmd = process.argv[2];

  if (cmd === 'list') {
    const r = await client.query('SELECT id, email, name FROM "user" ORDER BY email');
    console.log(JSON.stringify(
      r.rows.map((u) => ({ id: u.id, email: u.email, username: u.name, isAdmin: true }))
    ));

  } else if (cmd === 'reset') {
    const id = process.env.RESET_ID;
    const pwd = process.env.RESET_PWD;
    if (!id || !pwd) throw new Error('RESET_ID et RESET_PWD requis');

    const hash = await hashPassword(pwd); // scrypt better-auth, format attendu par l'app
    const up = await client.query(
      `UPDATE account SET password = $1, updated_at = now()
         WHERE user_id = $2 AND provider_id = 'credential'`,
      [hash, id]
    );

    // Vérification intégrée : on relit le hash stocké et on le valide avec le
    // verify natif de better-auth (la même fonction qu'au login).
    const row = await client.query(
      `SELECT password FROM account WHERE user_id = $1 AND provider_id = 'credential'`,
      [id]
    );
    const ok =
      up.rowCount === 1 &&
      row.rows[0] &&
      (await verifyPassword({ hash: row.rows[0].password, password: pwd }));
    console.log(ok ? 'OK' : 'FAIL');

  } else {
    throw new Error(`sous-commande inconnue: ${cmd}`);
  }
} finally {
  await client.end();
}

File: docuseal/accounts/ryvie-accounts.rb
# ryvie-accounts.rb — Fiche de gestion des comptes DocuSeal.
#
# Vit AVEC l'app (appstore) et s'exécute DANS le conteneur app-docuseal via
# `rails runner` : elle a donc un accès complet au modèle User (Devise) et à la
# base, sans dépendance externe. Ryvie ne fait que la déclencher (docker exec) et
# lire son résultat — il ne connaît ni le schéma ni le format de hash.
#
# DocuSeal hache via Devise (bcrypt). On réutilise Devise pour garantir la
# compatibilité au login.
#
# Sous-commandes (ARGV[0]) :
#   list                         -> JSON des comptes sur stdout
#   reset    (env RESET_ID/PWD)  -> change le mot de passe + vérifie -> "OK"/"FAIL"
#   verify   (env RESET_ID/PWD)  -> "OK" si le mot de passe correspond, sinon "NO"
#   provision (env DEFAULT_*)    -> crée le compte par défaut (idempotent) -> "DONE"
require 'json'

cmd = ARGV[0]

case cmd
when 'list'
  puts(User.all.map do |u|
    {
      id: u.id.to_s,
      email: u.email,
      isAdmin: (u.respond_to?(:admin?) ? !!u.admin? : (u.respond_to?(:role) ? u.role.to_s == 'admin' : false)),
    }
  end.to_json)

when 'reset'
  u = User.find(ENV['RESET_ID'])
  u.password = ENV['RESET_PWD']
  u.password_confirmation = ENV['RESET_PWD'] if u.respond_to?(:password_confirmation=)
  u.save!
  puts(u.reload.valid_password?(ENV['RESET_PWD']) ? 'OK' : 'FAIL')

when 'verify'
  u = (ENV['RESET_ID'].to_s.empty? ? nil : User.find_by(id: ENV['RESET_ID']))
  u ||= User.find_by(email: ENV['RESET_EMAIL']) unless ENV['RESET_EMAIL'].to_s.empty?
  puts(u && u.valid_password?(ENV['RESET_PWD']) ? 'OK' : 'NO')

when 'provision'
  email = ENV['DEFAULT_EMAIL']
  pwd = ENV['DEFAULT_PWD']
  uname = ENV['DEFAULT_USER']
  if User.find_by(email: email).nil?
    base = User.first
    acc = base ? base.account : (defined?(Account) ? Account.create!(name: 'Ryvie') : nil)
    u = User.new(email: email)
    u.first_name = uname if u.respond_to?(:first_name=)
    u.last_name = 'Ryvie' if u.respond_to?(:last_name=)
    u.role = (base ? base.role : 'admin') if u.respond_to?(:role=)
    (u.account = acc if acc) if u.respond_to?(:account=)
    u.password = pwd
    u.password_confirmation = pwd if u.respond_to?(:password_confirmation=)
    u.save!
  end
  puts 'DONE'

else
  warn "sous-commande inconnue: #{cmd}"
  exit 2
end
